### PR TITLE
adding prop and condition for OperationalInsightsEndpointResourceId

### PR DESCRIPTION
**Pull request by andrewpethel**
Task:
Update Azure.PowerShell.Common so that the Add-AzEnvironment auto discovery feature is agnostic for Sovereign Clouds.

Problem Description:
The cmdlet's current implementation relies on the ExtendedProperties setting the OperationalInsightsEndpoint on the active AzureEnvironment. For user-added environments, such as those added via Add-AzEnvironment, both the -OperationalInsightsEndpoint and -OperationalInsightsEndpointResourceID parameters must be defined for the Invoke command to work. This requirement is documented, but could be clearer and more user-friendly.

Proposed Solution:
This PR updates the MapArmToAzureEnvironment method to auto-populate the OperationalInsightsEndpoint and OperationalInsightsEndpointResourceId values, following a pattern similar to Synapse. This update would leverage the auto-discovery feature used with Add-AzEnvironment. By implementing this change, we can improve the robustness of the cmdlet and enhance the user experience for those working with custom Azure environments, namely Sovereign Clouds.

All tests pass (updated screenshot -- April 24, 2024):

image

### CODE REVIEW

File: src/Authentication.Abstractions/AzureEnvironment.cs [hunk 1]
```diff
@@ -231,6 +231,17 @@ private static AzureEnvironment MapArmToAzureEnvironment(ArmMetadata armMetadata
                 }
             }
 
+            if (!string.IsNullOrEmpty(armMetadata.LogAnalyticsResourceId))
+            {
+                string api_version = "v1";
+                var logAnalyticsEndpoint = armMetadata.LogAnalyticsResourceId + '/' + api_version;
+                azureEnvironment.SetProperty(ExtendedEndpoint.OperationalInsightsEndpointResourceId, armMetadata.LogAnalyticsResourceId);
+                if (!string.IsNullOrEmpty(logAnalyticsEndpoint))
```
Comment (Nickcandy) [inline src/Authentication.Abstractions/AzureEnvironment.cs:239]: In this case, logAnalyticsEndpoint would not be null or empty. This conditional statement could be removed.

Comment (andrewpethel) [reply src/Authentication.Abstractions/AzureEnvironment.cs]: Removing unnecessary conditional.

**Review by Nickcandy — APPROVED** [review]

**Review by msJinLei — DISMISSED** [review]
`OperationalInsightsEndpointResourceId` is already in extended property of `AzureEnvironment`. Please remove the new definition.

Please the change constant values of `OperationalInsightsEndpointResourceId` for all the environments
```
        public const string AzureOperationalInsightsEndpoint = "https://api.loganalytics.io/v1";
        public const string ChinaOperationalInsightsEndpoint = "https://api.loganalytics.azure.cn/v1";
        public const string USGovernmentOperationalInsightsEndpoint = "https://api.loganalytics.us/v1";
```

File: src/Authentication.Abstractions/AzureEnvironment.cs [hunk 2]
```diff
@@ -459,6 +467,11 @@ public AzureEnvironment(IAzureEnvironment other)
         /// </summary>
         public string ContainerRegistryEndpointSuffix { get; set; }
 
+        /// <summary>
+        /// The token audience required for communicating with the Log Analytics service in this enviornment
+        /// </summary>
+        public string OperationalInsightsEndpointResourceId { get; set; }
+
```
Comment (Nickcandy) [inline src/Authentication.Abstractions/AzureEnvironment.cs:474]: OperationalInsightsEndpointResourceId is defined in extendedProperties. I believe this defination is redundant.

File: src/Authentication.Abstractions/AzureEnvironment.cs [hunk 3]
```diff
@@ -231,6 +231,14 @@ private static AzureEnvironment MapArmToAzureEnvironment(ArmMetadata armMetadata
                 }
             }
 
+            if (!string.IsNullOrEmpty(armMetadata.LogAnalyticsResourceId))
+            {
+                string api_version = "v1";
+                var logAnalyticsEndpoint = armMetadata.LogAnalyticsResourceId + '/' + api_version;
+                azureEnvironment.SetProperty(ExtendedEndpoint.OperationalInsightsEndpointResourceId, armMetadata.LogAnalyticsResourceId);
+                azureEnvironment.SetProperty(ExtendedEndpoint.OperationalInsightsEndpoint, logAnalyticsEndpoint);
```
Comment (msJinLei) [inline src/Authentication.Abstractions/AzureEnvironment.cs:239]:
```suggestion
                azureEnvironment.SetProperty(ExtendedEndpoint.OperationalInsightsEndpoint, $"{armMetadata.LogAnalyticsResourceId}/v1");
```

File: src/Authentication.Abstractions/AzureEnvironment.cs [hunk 4]
```diff
@@ -231,6 +231,14 @@ private static AzureEnvironment MapArmToAzureEnvironment(ArmMetadata armMetadata
                 }
             }
 
+            if (!string.IsNullOrEmpty(armMetadata.LogAnalyticsResourceId))
+            {
+                string api_version = "v1";
```
Comment (msJinLei) [inline src/Authentication.Abstractions/AzureEnvironment.cs:236]:
```suggestion

```

File: src/Authentication.Abstractions/AzureEnvironment.cs [hunk 5]
```diff
@@ -231,6 +231,14 @@ private static AzureEnvironment MapArmToAzureEnvironment(ArmMetadata armMetadata
...
+                var logAnalyticsEndpoint = armMetadata.LogAnalyticsResourceId + '/' + api_version;
```
Comment (msJinLei) [inline src/Authentication.Abstractions/AzureEnvironment.cs:237]:
```suggestion

```

File: src/Authentication.Abstractions/AzureEnvironment.cs [hunk 6]
```diff
@@ -459,6 +467,11 @@ public AzureEnvironment(IAzureEnvironment other)
         /// </summary>
         public string ContainerRegistryEndpointSuffix { get; set; }
 
+        /// <summary>
+        /// The token audience required for communicating with the Log Analytics service in this enviornment
+        /// </summary>
+        public string OperationalInsightsEndpointResourceId { get; set; }
```
Comment (msJinLei) [inline src/Authentication.Abstractions/AzureEnvironment.cs:473]:
```suggestion

```

**Review by msJinLei — CHANGES_REQUESTED** [review]

File: src/Authentication.Abstractions/AzureEnvironment.cs (hunk 2 above)
Comment (andrewpethel) [reply src/Authentication.Abstractions/AzureEnvironment.cs:474]: Yes, I agree. Removing redundancy.

File: src/Authentication.Abstractions/AzureEnvironment.cs (hunk 6 above)
Comment (andrewpethel) [reply src/Authentication.Abstractions/AzureEnvironment.cs:473]: removed.

File: src/Authentication.Abstractions/AzureEnvironment.cs (hunk 3 above)
Comment (andrewpethel) [reply src/Authentication.Abstractions/AzureEnvironment.cs]: simplified.

File: src/Authentication.Abstractions/AzureEnvironment.cs (hunk 4 above)
Comment (andrewpethel) [reply src/Authentication.Abstractions/AzureEnvironment.cs]: removed

File: src/Authentication.Abstractions/AzureEnvironment.cs (hunk 5 above)
Comment (andrewpethel) [reply src/Authentication.Abstractions/AzureEnvironment.cs]: removed.

**Review by msJinLei — APPROVED** [review]

**Review by isra-fel — APPROVED** [review]
@andrewpethel thanks a lot for the contribution! This will be included in the upcoming milestone.

**Comment by andrewpethel** [comment]
[heart]         Ryan Pethel reacted to your message:
________________________________

**Pull request merged**